File: java/Benchmark.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.management.ManagementFactory;
import java.lang.management.ThreadMXBean;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import sort.SelectionSort;
import sort.MergeSort;
import sort.BubbleSort;
import sort.HeapSort;

public class Benchmark {

    @FunctionalInterface
    interface AlgorithmFunction {
        int[] run(int[] arr);
    }

    // null -> run every algorithm. A set -> run only those algo names.
    private static Set<String> selectedAlgos = null;

    public static void main(String[] args) {

        selectedAlgos = parseSelectedAlgos(args);
        if (selectedAlgos != null) {
            System.out.println("Running only: " + selectedAlgos);
        }

        File randomResultsDir =
                new File("../results/java/random");

        File duplicatesResultsDir =
                new File("../results/java/duplicates");

        File bigDataResultsDir =
                new File("../results/java/bigdata");

        createDirectory(randomResultsDir);
        createDirectory(duplicatesResultsDir);
        createDirectory(bigDataResultsDir);

        String[] allAlgoNames = {
                "SelectionSort",
                "ImprovedSelectionSort",
                "RecursiveMergeSort",
                "IterativeMergeSort",
                "BubbleSort",
                "BasicBubbleSort",
                "HeapSort",
                "RecursiveHeapSort"
        };

        String[] algoNames = filterAlgoNames(allAlgoNames);

        String randomFilePath =
                "../data/randomArrays.txt";

        System.out.println();
        System.out.println("############################################");
        System.out.println("#           RANDOM DATA TESTING            #");
        System.out.println("#                  3 RUNS                   #");
        System.out.println("############################################");

        prepareResultFiles(
                randomResultsDir,
                algoNames
        );

        processInputFile(
                randomFilePath,
                randomResultsDir,
                3
        );

        String[] duplicateFiles = {
                "../data/duplicates_1000.csv",
                "../data/duplicates_5000.csv",
                "../data/duplicates_10000.csv",
                "../data/duplicates_50000.csv"
        };

        System.out.println();
        System.out.println("############################################");
        System.out.println("#         DUPLICATES DATA TESTING          #");
        System.out.println("#                 30 RUNS                  #");
        System.out.println("############################################");

        prepareResultFiles(
                duplicatesResultsDir,
                algoNames
        );

        for (String duplicateFile : duplicateFiles) {

            System.out.println();
            System.out.println("============================================");
            System.out.println(
                    "TESTING FILE: " + duplicateFile
            );
            System.out.println("RUNS: 30");
            System.out.println("============================================");

            processInputFile(
                    duplicateFile,
                    duplicatesResultsDir,
                    30
            );
        }

        String[] bigDataFiles = {
                "../data/big_datasets_unique_1_100000.txt",
                "../data/big_datasets_unique_2_100000.txt"
        };

        System.out.println();
        System.out.println("############################################");
        System.out.println("#            BIG DATA TESTING              #");
        System.out.println("#                  30 RUNS                 #");
        System.out.println("#        EACH FILE = ONE BIG ARRAY         #");
        System.out.println("############################################");

        prepareResultFiles(
                bigDataResultsDir,
                algoNames
        );

        for (String bigDataFile : bigDataFiles) {

            System.out.println();
            System.out.println("============================================");
            System.out.println(
                    "TESTING BIG DATA FILE: " + bigDataFile
            );
            System.out.println("RUNS: 30");
            System.out.println("============================================");

            try {

                processBigDataFile(
                        bigDataFile,
                        bigDataResultsDir,
                        30
                );

            } catch (IOException e) {

                System.err.println(
                        "Greska pri citanju big data fajla: "
                                + bigDataFile
                );

                System.err.println(
                        e.getMessage()
                );
            }
        }

        System.out.println();
        System.out.println("############################################");
        System.out.println("#            TESTING FINISHED              #");
        System.out.println("############################################");

        System.out.println();
        System.out.println("Rezultati:");

        System.out.println(
                "../results/java/random/"
        );

        System.out.println(
                "../results/java/duplicates/"
        );

        System.out.println(
                "../results/java/bigdata/"
        );
    }

    /**
     * Reads the algorithm filter from the command line.
     * Returns null when no argument was given (= run everything).
     * Accepts full names ("BubbleSort") or shortcuts ("bubble", "heap",
     * "selection", "merge"), separated by spaces or commas.
     */
    private static Set<String> parseSelectedAlgos(String[] args) {

        if (args == null || args.length == 0) {
            return null;
        }

        Set<String> selected = new HashSet<>();

        for (String arg : args) {

            for (String token : arg.split("[,\\s]+")) {

                if (token.isEmpty()) {
                    continue;
                }

                switch (token.toLowerCase()) {
                    case "bubble":
                        selected.add("BubbleSort");
                        selected.add("BasicBubbleSort");
                        break;
                    case "heap":
                        selected.add("HeapSort");
                        selected.add("RecursiveHeapSort");
                        break;
                    case "selection":
                        selected.add("SelectionSort");
                        selected.add("ImprovedSelectionSort");
                        break;
                    case "merge":
                        selected.add("RecursiveMergeSort");
                        selected.add("IterativeMergeSort");
                        break;
                    default:
                        selected.add(token);
                }
            }
        }

        return selected;
    }

    /** Keeps only the algorithm names that pass the command-line filter. */
    private static String[] filterAlgoNames(String[] allAlgoNames) {

        if (selectedAlgos == null) {
            return allAlgoNames;
        }

        List<String> kept = new ArrayList<>();

        for (String name : allAlgoNames) {
            if (selectedAlgos.contains(name)) {
                kept.add(name);
            }
        }

        return kept.toArray(new String[0]);
    }

    private static void createDirectory(File directory) {

        if (!directory.exists()) {

            if (directory.mkdirs()) {

                System.out.println(
                        "Created directory: "
                                + directory.getPath()
                );

            } else {

                System.err.println(
                        "Could not create directory: "
                                + directory.getPath()
                );
            }
        }
    }

    private static void prepareResultFiles(
            File resultsDir,
            String[] algoNames) {

        for (String name : algoNames) {

            File resultFile =
                    new File(
                            resultsDir,
                            name + "Result.csv"
                    );

            if (resultFile.exists()) {

                if (!resultFile.delete()) {

                    System.err.println(
                            "Could not delete old result file: "
                                    + resultFile.getPath()
                    );
                }
            }

            try (BufferedWriter writer =
                         new BufferedWriter(
                                 new FileWriter(
                                         resultFile,
                                         true
                                 )
                         )) {

                writer.write(
                        "dataset,run,arraylength,time,cputime,memory"
                );

                writer.newLine();

            } catch (IOException e) {

                System.err.println(
                        "Greska pri kreiranju fajla za "
                                + name
                                + ": "
                                + e.getMessage()
                );
            }
        }
    }

    private static void processInputFile(
            String filePath,
            File resultsDir,
            int numberOfRuns) {

        String datasetName =
                new File(filePath).getName();

        int dotIndex =
                datasetName.lastIndexOf('.');

        if (dotIndex > 0) {

            datasetName =
                    datasetName.substring(
                            0,
                            dotIndex
                    );
        }

        for (int run = 1; run <= numberOfRuns; run++) {

            System.out.println();
            System.out.println(
                    "============================================"
            );

            System.out.println(
                    "Dataset: " + datasetName
            );

            System.out.println(
                    "Run: " + run + "/" + numberOfRuns
            );

            System.out.println(
                    "============================================"
            );

            try {

                List<int[]> arrays =
                        readArraysFromFile(filePath);

                int arrayCounter = 1;

                for (int[] originalArr : arrays) {

                    System.out.println(
                            "=== RUN #" + run
                                    + " | ARRAY #" + arrayCounter
                                    + " | Length: "
                                    + originalArr.length
                                    + " ==="
                    );

                    measureAndSave(
                            datasetName,
                            run,
                            "SelectionSort",
                            originalArr,
                            resultsDir,
                            arr -> {

                                SelectionSort.selection_sort(arr);

                                return arr;
                            }
                    );

                    measureAndSave(
                            datasetName,
                            run,
                            "ImprovedSelectionSort",
                            originalArr,
                            resultsDir,
                            arr -> {

                                SelectionSort.improved_selection_sort(arr);

                                return arr;
                            }
                    );

                    measureAndSave(
                            datasetName,
                            run,
                            "RecursiveMergeSort",
                            originalArr,
                            resultsDir,
                            arr ->
                                    MergeSort.mergeSort(arr)
                    );

                    measureAndSave(
                            datasetName,
                            run,
                            "IterativeMergeSort",
                            originalArr,
                            resultsDir,
                            arr ->
                                    MergeSort.mergeSortWithoutRecursion(arr)
                    );

                    measureAndSave(
                            datasetName,
                            run,
                            "BubbleSort",
                            originalArr,
                            resultsDir,
                            arr -> {

                                BubbleSort.sort(arr);

                                return arr;
                            }
                    );

                    measureAndSave(
                            datasetName,
                            run,
                            "BasicBubbleSort",
                            originalArr,
                            resultsDir,
                            arr -> {

                                BubbleSort.sortBasic(arr);

                                return arr;
                            }
                    );

                    measureAndSave(
                            datasetName,
                            run,
                            "HeapSort",
                            originalArr,
                            resultsDir,
                            arr -> {

                                HeapSort.sort(arr);

                                return arr;
                            }
                    );

                    measureAndSave(
                            datasetName,
                            run,
                            "RecursiveHeapSort",
                            originalArr,
                            resultsDir,
                            arr -> {

                                HeapSort.sortRecursive(arr);

                                return arr;
                            }
                    );

                    System.out.println(
                            "--------------------------------------------"
                    );

                    arrayCounter++;
                }

            } catch (IOException e) {

                System.out.println(
                        "Error: The file '"
                                + filePath
                                + "' does not exist!"
                );

                System.out.println(
                        "Please check the data directory."
                );
            }
        }
    }

    private static void processBigDataFile(
            String filePath,
            File resultsDir,
            int numberOfRuns) throws IOException {

        String datasetName =
                new File(filePath).getName();

        int dotIndex =
                datasetName.lastIndexOf('.');

        if (dotIndex > 0) {

            datasetName =
                    datasetName.substring(
                            0,
                            dotIndex
                    );
        }

        int[] originalArr =
                readBigDataFile(filePath);

        System.out.println();
        System.out.println(
                "Dataset: " + datasetName
        );

        System.out.println(
                "Array length: "
                        + originalArr.length
        );

        for (int run = 1; run <= numberOfRuns; run++) {

            System.out.println();
            System.out.println(
                    "============================================"
            );

            System.out.println(
                    "Dataset: " + datasetName
            );

            System.out.println(
                    "Run: " + run + "/" + numberOfRuns
            );

            System.out.println(
                    "Array length: "
                            + originalArr.length
            );

            System.out.println(
                    "============================================"
            );

            measureAndSave(
                    datasetName,
                    run,
                    "SelectionSort",
                    originalArr,
                    resultsDir,
                    arr -> {

                        SelectionSort.selection_sort(arr);

                        return arr;
                    }
            );

            measureAndSave(
                    datasetName,
                    run,
                    "ImprovedSelectionSort",
                    originalArr,
                    resultsDir,
                    arr -> {

                        SelectionSort.improved_selection_sort(arr);

                        return arr;
                    }
            );

            measureAndSave(
                    datasetName,
                    run,
                    "RecursiveMergeSort",
                    originalArr,
                    resultsDir,
                    arr ->
                            MergeSort.mergeSort(arr)
            );

            measureAndSave(
                    datasetName,
                    run,
                    "IterativeMergeSort",
                    originalArr,
                    resultsDir,
                    arr ->
                            MergeSort.mergeSortWithoutRecursion(arr)
            );

            measureAndSave(
                    datasetName,
                    run,
                    "BubbleSort",
                    originalArr,
                    resultsDir,
                    arr -> {

                        BubbleSort.sort(arr);

                        return arr;
                    }
            );

            measureAndSave(
                    datasetName,
                    run,
                    "BasicBubbleSort",
                    originalArr,
                    resultsDir,
                    arr -> {

                        BubbleSort.sortBasic(arr);

                        return arr;
                    }
            );

            measureAndSave(
                    datasetName,
                    run,
                    "HeapSort",
                    originalArr,
                    resultsDir,
                    arr -> {

                        HeapSort.sort(arr);

                        return arr;
                    }
            );

            measureAndSave(
                    datasetName,
                    run,
                    "RecursiveHeapSort",
                    originalArr,
                    resultsDir,
                    arr -> {

                        HeapSort.sortRecursive(arr);

                        return arr;
                    }
            );

            System.out.println(
                    "--------------------------------------------"
            );
        }
    }

    private static List<int[]> readArraysFromFile(
            String filePath) throws IOException {

        List<int[]> arrays =
                new ArrayList<>();

        try (BufferedReader br =
                     new BufferedReader(
                             new FileReader(filePath)
                     )) {

            String line;

            while ((line = br.readLine()) != null) {

                line = line.trim();

                if (line.isEmpty()) {
                    continue;
                }

                String[] tokens =
                        line.split("[,\\s]+");

                int[] arr =
                        new int[tokens.length];

                boolean validLine = true;

                for (int i = 0; i < tokens.length; i++) {

                    try {

                        arr[i] =
                                Integer.parseInt(
                                        tokens[i].trim()
                                );

                    } catch (NumberFormatException e) {

                        validLine = false;

                        System.err.println(
                                "Preskacem nevalidan red: "
                                        + line
                        );

                        break;
                    }
                }

                if (validLine && arr.length > 0) {

                    arrays.add(arr);
                }
            }
        }

        return arrays;
    }

    private static int[] readBigDataFile(
            String filePath) throws IOException {

        List<Integer> numbers =
                new ArrayList<>();

        try (BufferedReader br =
                     new BufferedReader(
                             new FileReader(filePath)
                     )) {

            String line;

            while ((line = br.readLine()) != null) {

                line = line.trim();

                if (line.isEmpty()) {
                    continue;
                }

                String[] tokens =
                        line.split("[,\\s]+");

                for (String token : tokens) {

                    if (!token.isEmpty()) {

                        numbers.add(
                                Integer.parseInt(
                                        token.trim()
                                )
                        );
                    }
                }
            }
        }

        int[] array =
                new int[numbers.size()];

        for (int i = 0; i < numbers.size(); i++) {

            array[i] =
                    numbers.get(i);
        }

        return array;
    }

    private static void measureAndSave(
            String datasetName,
            int run,
            String algoName,
            int[] originalArr,
            File resultsDir,
            AlgorithmFunction algo) {

        if (selectedAlgos != null && !selectedAlgos.contains(algoName)) {
            return;
        }

        int[] arrCopy =
                Arrays.copyOf(
                        originalArr,
                        originalArr.length
                );

        ThreadMXBean threadBean =
                ManagementFactory.getThreadMXBean();

        System.gc();

        try {

            Thread.sleep(10);

        } catch (InterruptedException ignored) {
        }

        Runtime runtime =
                Runtime.getRuntime();

        long memBefore =
                runtime.totalMemory()
                        - runtime.freeMemory();

        long startWallTime =
                System.nanoTime();

        long startCpuTime =
                threadBean.isCurrentThreadCpuTimeSupported()
                        ? threadBean.getCurrentThreadCpuTime()
                        : 0;

        int[] sortedArr =
                algo.run(arrCopy);

        long endCpuTime =
                threadBean.isCurrentThreadCpuTimeSupported()
                        ? threadBean.getCurrentThreadCpuTime()
                        : 0;

        long endWallTime =
                System.nanoTime();

        long memAfter =
                runtime.totalMemory()
                        - runtime.freeMemory();

        double wallTimeMs =
                (endWallTime - startWallTime)
                        / 1_000_000.0;

        double cpuTimeMs =
                (endCpuTime - startCpuTime)
                        / 1_000_000.0;

        double memoryUsedKb =
                Math.max(
                        0,
                        (memAfter - memBefore)
                                / 1024.0
                );

        System.out.print(
                "  [" + algoName + "] (first 10): "
        );

        printFirstTen(sortedArr);

        System.out.printf(
                "    -> Time: %.2f ms"
                        + " | CPU Time: %.2f ms"
                        + " | Memory: %.2f KB%n",
                wallTimeMs,
                cpuTimeMs,
                memoryUsedKb
        );

        File resultFile =
                new File(
                        resultsDir,
                        algoName + "Result.csv"
                );

        try (BufferedWriter writer =
                     new BufferedWriter(
                             new FileWriter(
                                     resultFile,
                                     true
                             )
                     )) {

            writer.write(
                    String.format(
                            "%s,%d,%d,%.2f,%.2f,%.2f",
                            datasetName,
                            run,
                            originalArr.length,
                            wallTimeMs,
                            cpuTimeMs,
                            memoryUsedKb
                    )
            );

            writer.newLine();

        } catch (IOException e) {

            System.err.println(
                    "Greska pri upisu rezultata za "
                            + algoName
                            + ": "
                            + e.getMessage()
            );
        }
    }

    private static void printFirstTen(int[] arr) {

        int limit =
                Math.min(
                        10,
                        arr.length
                );

        int[] firstTen =
                Arrays.copyOfRange(
                        arr,
                        0,
                        limit
                );

        System.out.println(
                Arrays.toString(firstTen)
        );
    }
}

File: python/Benchmark.py
import csv
import os
import sys
import time
import psutil

from sort.SelectionSort import SelectionSort
from sort.MergeSort import MergeSort
from sort.BubbleSort import BubbleSort
from sort.HeapSort import HeapSort


# None -> run every algorithm. A set -> run only those algo_name values.
SELECTED_ALGOS = None

# Friendly shortcuts you can pass on the command line instead of full names.
ALGO_ALIASES = {
    "bubble": ["BubbleSort", "BasicBubbleSort"],
    "heap": ["HeapSort", "RecursiveHeapSort"],
    "selection": ["SelectionSort", "ImprovedSelectionSort"],
    "merge": ["RecursiveMergeSort", "IterativeMergeSort"],
}


def parse_selected_algos(args):
    if not args:
        return None
    tokens = " ".join(args).replace(",", " ").split()
    selected = set()
    for token in tokens:
        key = token.lower()
        if key in ALGO_ALIASES:
            selected.update(ALGO_ALIASES[key])
        else:
            selected.add(token)
    return selected


def main():

    global SELECTED_ALGOS
    SELECTED_ALGOS = parse_selected_algos(sys.argv[1:])
    if SELECTED_ALGOS is not None:
        print("Running only: " + ", ".join(sorted(SELECTED_ALGOS)))

    random_results_dir = "../results/python/random"
    duplicates_results_dir = "../results/python/duplicates"
    big_data_results_dir = "../results/python/bigdata"

    create_directory(random_results_dir)
    create_directory(duplicates_results_dir)
    create_directory(big_data_results_dir)

    algo_names = [
        "SelectionSort",
        "ImprovedSelectionSort",
        "RecursiveMergeSort",
        "IterativeMergeSort",
        "BubbleSort",
        "BasicBubbleSort",
        "HeapSort",
        "RecursiveHeapSort"
    ]

    if SELECTED_ALGOS is not None:
        algo_names = [
            name for name in algo_names if name in SELECTED_ALGOS
        ]

    random_file_path = "../data/randomArrays.txt"

    print()
    print("############################################")
    print("#           RANDOM DATA TESTING            #")
    print("#                  3 RUNS                   #")
    print("############################################")

    prepare_result_files(
        random_results_dir,
        algo_names
    )

    process_input_file(
        random_file_path,
        random_results_dir,
        3
    )

    duplicate_files = [
        "../data/duplicates_1000.csv",
        "../data/duplicates_5000.csv",
        "../data/duplicates_10000.csv",
        "../data/duplicates_50000.csv"
    ]

    print()
    print("############################################")
    print("#         DUPLICATES DATA TESTING          #")
    print("#        30 RUNS / 50000 = 2 RUNS          #")
    print("############################################")

    prepare_result_files(
        duplicates_results_dir,
        algo_names
    )

    for duplicate_file in duplicate_files:

        if "duplicates_50000" in duplicate_file:
            number_of_runs = 2
        else:
            number_of_runs = 30

        print()
        print("============================================")
        print("TESTING FILE: " + duplicate_file)
        print("RUNS: " + str(number_of_runs))
        print("============================================")

        process_input_file(
            duplicate_file,
            duplicates_results_dir,
            number_of_runs
        )

    big_data_files = [
        "../data/big_datasets_unique_1_100000.txt",
        "../data/big_datasets_unique_2_100000.txt"
    ]

    print()
    print("############################################")
    print("#            BIG DATA TESTING              #")
    print("#                  1 RUN                   #")
    print("#        EACH FILE = ONE BIG ARRAY         #")
    print("############################################")

    prepare_result_files(
        big_data_results_dir,
        algo_names
    )

    for big_data_file in big_data_files:

        print()
        print("============================================")
        print("TESTING BIG DATA FILE: " + big_data_file)
        print("RUNS: 1")
        print("============================================")

        try:

            process_big_data_file(
                big_data_file,
                big_data_results_dir,
                1
            )

        except (OSError, ValueError) as e:

            print(
                "Greska pri citanju big data fajla: "
                + big_data_file
            )

            print(str(e))

    print()
    print("############################################")
    print("#            TESTING FINISHED              #")
    print("############################################")

    print()
    print("Rezultati:")

    print("../results/python/random/")
    print("../results/python/duplicates/")
    print("../results/python/bigdata/")


def create_directory(directory):

    if not os.path.exists(directory):

        try:

            os.makedirs(directory)

            print(
                "Created directory: "
                + directory
            )

        except OSError:

            print(
                "Could not create directory: "
                + directory
            )


def prepare_result_files(results_dir, algo_names):

    for name in algo_names:

        result_file = os.path.join(
            results_dir,
            name + "Result.csv"
        )

        if os.path.exists(result_file):

            try:

                os.remove(result_file)

            except OSError:

                print(
                    "Could not delete old result file: "
                    + result_file
                )

        try:

            with open(
                result_file,
                "w",
                newline="",
                encoding="utf-8"
            ) as file:

                writer = csv.writer(file)

                writer.writerow([
                    "dataset",
                    "run",
                    "arraylength",
                    "time",
                    "cputime",
                    "memory"
                ])

        except OSError as e:

            print(
                "Greska pri kreiranju fajla za "
                + name
                + ": "
                + str(e)
            )


def process_input_file(
        file_path,
        results_dir,
        number_of_runs):

    dataset_name = os.path.basename(file_path)

    dataset_name = os.path.splitext(
        dataset_name
    )[0]

    for run in range(1, number_of_runs + 1):

        print()
        print(
            "============================================"
        )

        print(
            "Dataset: " + dataset_name
        )

        print(
            "Run: "
            + str(run)
            + "/"
            + str(number_of_runs)
        )

        print(
            "============================================"
        )

        try:

            arrays = read_arrays_from_file(
                file_path
            )

            array_counter = 1

            for original_arr in arrays:

                print(
                    "=== RUN #"
                    + str(run)
                    + " | ARRAY #"
                    + str(array_counter)
                    + " | Length: "
                    + str(len(original_arr))
                    + " ==="
                )

                measure_and_save(
                    dataset_name,
                    run,
                    "SelectionSort",
                    original_arr,
                    results_dir,
                    SelectionSort.selection_sort
                )

                measure_and_save(
                    dataset_name,
                    run,
                    "ImprovedSelectionSort",
                    original_arr,
                    results_dir,
                    SelectionSort.improved_selection_sort
                )

                measure_and_save(
                    dataset_name,
                    run,
                    "RecursiveMergeSort",
                    original_arr,
                    results_dir,
                    MergeSort.mergeSort
                )

                measure_and_save(
                    dataset_name,
                    run,
                    "IterativeMergeSort",
                    original_arr,
                    results_dir,
                    MergeSort.mergeSortWithoutRecursion
                )

                measure_and_save(
                    dataset_name,
                    run,
                    "BubbleSort",
                    original_arr,
                    results_dir,
                    BubbleSort.sort
                )

                measure_and_save(
                    dataset_name,
                    run,
                    "BasicBubbleSort",
                    original_arr,
                    results_dir,
                    BubbleSort.sort_basic
                )

                measure_and_save(
                    dataset_name,
                    run,
                    "HeapSort",
                    original_arr,
                    results_dir,
                    HeapSort.sort
                )

                measure_and_save(
                    dataset_name,
                    run,
                    "RecursiveHeapSort",
                    original_arr,
                    results_dir,
                    HeapSort.sort_recursive
                )

                print(
                    "--------------------------------------------"
                )

                array_counter += 1

        except OSError:

            print(
                "Error: The file '"
                + file_path
                + "' does not exist!"
            )

            print(
                "Please check the data directory."
            )


def process_big_data_file(
        file_path,
        results_dir,
        number_of_runs):

    dataset_name = os.path.basename(file_path)

    dataset_name = os.path.splitext(
        dataset_name
    )[0]

    original_arr = read_big_data_file(
        file_path
    )

    print()
    print(
        "Dataset: " + dataset_name
    )

    print(
        "Array length: "
        + str(len(original_arr))
    )

    for run in range(1, number_of_runs + 1):

        print()
        print(
            "============================================"
        )

        print(
            "Dataset: " + dataset_name
        )

        print(
            "Run: "
            + str(run)
            + "/"
            + str(number_of_runs)
        )

        print(
            "Array length: "
            + str(len(original_arr))
        )

        print(
            "============================================"
        )

        measure_and_save(
            dataset_name,
            run,
            "SelectionSort",
            original_arr,
            results_dir,
            SelectionSort.selection_sort
        )

        measure_and_save(
            dataset_name,
            run,
            "ImprovedSelectionSort",
            original_arr,
            results_dir,
            SelectionSort.improved_selection_sort
        )

        measure_and_save(
            dataset_name,
            run,
            "RecursiveMergeSort",
            original_arr,
            results_dir,
            MergeSort.mergeSort
        )

        measure_and_save(
            dataset_name,
            run,
            "IterativeMergeSort",
            original_arr,
            results_dir,
            MergeSort.mergeSortWithoutRecursion
        )

        measure_and_save(
            dataset_name,
            run,
            "BubbleSort",
            original_arr,
            results_dir,
            BubbleSort.sort
        )

        measure_and_save(
            dataset_name,
            run,
            "BasicBubbleSort",
            original_arr,
            results_dir,
            BubbleSort.sort_basic
        )

        measure_and_save(
            dataset_name,
            run,
            "HeapSort",
            original_arr,
            results_dir,
            HeapSort.sort
        )

        measure_and_save(
            dataset_name,
            run,
            "RecursiveHeapSort",
            original_arr,
            results_dir,
            HeapSort.sort_recursive
        )

        print(
            "--------------------------------------------"
        )


def read_arrays_from_file(file_path):

    arrays = []

    with open(
        file_path,
        "r",
        encoding="utf-8"
    ) as file:

        for line in file:

            line = line.strip()

            if not line:
                continue

            tokens = line.replace(
                ",",
                " "
            ).split()

            arr = []

            valid_line = True

            for token in tokens:

                try:

                    arr.append(
                        int(token.strip())
                    )

                except ValueError:

                    valid_line = False

                    print(
                        "Preskacem nevalidan red: "
                        + line
                    )

                    break

            if valid_line and len(arr) > 0:

                arrays.append(arr)

    return arrays


def read_big_data_file(file_path):

    numbers = []

    with open(
        file_path,
        "r",
        encoding="utf-8"
    ) as file:

        for line in file:

            line = line.strip()

            if not line:
                continue

            tokens = line.replace(
                ",",
                " "
            ).split()

            for token in tokens:

                if token:

                    numbers.append(
                        int(token.strip())
                    )

    return numbers


def measure_and_save(
        dataset_name,
        run,
        algo_name,
        original_arr,
        results_dir,
        algorithm):

    if SELECTED_ALGOS is not None and algo_name not in SELECTED_ALGOS:
        return

    arr_copy = original_arr.copy()

    process = psutil.Process(
        os.getpid()
    )

    time.sleep(0.01)

    memory_before = (
        process.memory_info().rss
    )

    start_wall_time = (
        time.perf_counter()
    )

    start_cpu_time = (
        time.process_time()
    )

    sorted_arr = algorithm(
        arr_copy
    )

    end_cpu_time = (
        time.process_time()
    )

    end_wall_time = (
        time.perf_counter()
    )

    memory_after = (
        process.memory_info().rss
    )

    if sorted_arr is None:

        sorted_arr = arr_copy

    wall_time_ms = (
        end_wall_time
        - start_wall_time
    ) * 1000.0

    cpu_time_ms = (
        end_cpu_time
        - start_cpu_time
    ) * 1000.0

    memory_used_kb = max(
        0,
        (
            memory_after
            - memory_before
        ) / 1024.0
    )

    print(
        "  ["
        + algo_name
        + "] (first 10): ",
        end=""
    )

    print_first_ten(
        sorted_arr
    )

    print(
        "    -> Time: "
        + format(
            wall_time_ms,
            ".2f"
        )
        + " ms"
        + " | CPU Time: "
        + format(
            cpu_time_ms,
            ".2f"
        )
        + " ms"
        + " | Memory: "
        + format(
            memory_used_kb,
            ".2f"
        )
        + " KB"
    )

    result_file = os.path.join(
        results_dir,
        algo_name + "Result.csv"
    )

    try:

        with open(
            result_file,
            "a",
            newline="",
            encoding="utf-8"
        ) as file:

            writer = csv.writer(file)

            writer.writerow([
                dataset_name,
                run,
                len(original_arr),
                format(
                    wall_time_ms,
                    ".2f"
                ),
                format(
                    cpu_time_ms,
                    ".2f"
                ),
                format(
                    memory_used_kb,
                    ".2f"
                )
            ])

    except OSError as e:

        print(
            "Greska pri upisu rezultata za "
            + algo_name
            + ": "
            + str(e)
        )


def print_first_ten(arr):

    limit = min(
        10,
        len(arr)
    )

    first_ten = arr[:limit]

    print(first_ten)


if __name__ == "__main__":
    main()
